clients/drcachesim/tests/invariant_checker_test.cpp

```diff
@@ -355,6 +355,207 @@ check_sane_control_flow()
+    {
+        std::vector<memref_t> memrefs = {
+            gen_instr(TID, 101),
+            // First signal.
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_EVENT, 102),
+            gen_instr(TID, 201),
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_XFER, 202),
+            // Second signal.
+            // The Marker value for this signal needs to be 102.
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_EVENT, 102),
+            gen_instr(TID, 201),
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_XFER, 202),
+            gen_instr(TID, 102),
+        };
+        if (!run_checker(memrefs, false)) {
+            return false;
+        }
+    }
+    // Incorrect test: back-to-back signals without any intervening instruction.
+    {
+        std::vector<memref_t> memrefs = {
+            gen_instr(TID, 101),
+            // First signal.
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_EVENT, 102),
+            gen_instr(TID, 201),
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_XFER, 202),
+            // Second signal.
+            // There will be a PC discontinuity here since the marker value is 500, and
+            // the previous PC is 101.
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_EVENT, 500),
+            gen_instr(TID, 201),
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_XFER, 202),
+        };
+        if (!run_checker(
+                memrefs, true,
+                { "Non-explicit control flow has no marker @ kernel_event marker", TID,
+                  /*ref_ordinal=*/5, /*last_timestamp=*/0,
+                  /*instrs_since_last_timestamp=*/2 },
+                "Failed to catch PC discontinuity for back-to-back signals without any "
+                "intervening instruction")) {
+            return false;
+        }
+    }
+    // Correct test: Taken branch with signal in between branch and its target.
+    {
+        instr_t *move = XINST_CREATE_move(GLOBAL_DCONTEXT, opnd_create_reg(REG1),
+                                          opnd_create_reg(REG2));
+        instr_t *cbr_to_move =
+            XINST_CREATE_jump_cond(GLOBAL_DCONTEXT, DR_PRED_EQ, opnd_create_instr(move));
+        instr_t *nop = XINST_CREATE_nop(GLOBAL_DCONTEXT);
+        instrlist_t *ilist = instrlist_create(GLOBAL_DCONTEXT);
+        instrlist_append(ilist, cbr_to_move);
+        instrlist_append(ilist, nop);
+        instrlist_append(ilist, move);
+        static constexpr addr_t BASE_ADDR = 0x123450;
+        std::vector<memref_with_IR_t> memref_setup = {
+            { gen_marker(TID, TRACE_MARKER_TYPE_VERSION, TRACE_ENTRY_VERSION_BRANCH_INFO),
+              nullptr },
+            { gen_marker(TID, TRACE_MARKER_TYPE_FILETYPE, OFFLINE_FILE_TYPE_ENCODINGS),
+              nullptr },
+            { gen_instr_type(TRACE_TYPE_INSTR_TAKEN_JUMP, TID), cbr_to_move },
+            { gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_EVENT, 0), move },
```

Comment:
OK so add_encodings_to_memrefs() replaces the 0 marker value with the PC of `move`.  How about `s/0/kWillBeReplaced/` or a comment by the 0 to avoid future readers being confused?